---
title: "BCB420.2019.GWAS"
author: "Sapir Labes"
date: "3 בפברואר 2019"
output: html_document
---
# `BCB420.2019.GWAS`

#### (**GWAS data annotatation of human genes**)


###### [Sapir  Labes](https://orcid.org/0000-0002-9478-5974), Department of Molecular Biology, Immunology and Cancer Research, The Hebrew University of Jerusalem, Israel. &lt;sapir.labes@mail.huji.ac.il&gt;
###### The template for this README file, as well as most written under "about this package", was quoted and paraphrased from the author Boris Steipe, [the link](https://github.com/hyginn/BCB420.2019.STRING)

----

**If any of this information is ambiguous, inaccurate, outdated, or incomplete, please check the [most recent version](https://github.com/sapirl/BCB420.2019.GWAS) of the package on GitHub and [file an issue](https://github.com/sapirl/BCB420.2019.GWAS/issues).**

----

### 1 About this package:

----

This package describes the workflow to download SNPs data and their association with human diseases from [the GWAS database](https://www.ebi.ac.uk/gwas/docs/file-downloads), how to map the IDs to [HGNC](https://www.genenames.org/) symbols, how to annotate the example gene set, and provides examples of computing database statistics.
<br>

The package serves dual duty, as an RStudio project, as well as an R package that can be installed. Package checks pass without errors, warnings, or notes.
<br>

This package supports R users, researchers and developers that want to map or index the GWAS data to  HGNC symbols.
<br>

or

<br>

This package is a mapping tool between GWAS database and HGNC gene symbols. It includes the description of each of the needed steps for performing the mapping, as well as the required packages and codes.<br>
This package contains descriptions and codes for downloading the database, mapping it, and annotating the example gene set.<br>
<br>

**In this project**
<br>

```
 --BCB420.2019.GWAS/
   |__.gitignore
   |__.Rbuildignore
   |__BCB420.2019.GWAS.Rproj
   |__DESCRIPTION
   |__dev/
      |__rptTwee.R
      |__toBrowser.R               # display .md files in your browser
   |__inst/
      |__extdata/
      |__scripts/
   |__LICENSE
   |__NAMESPACE
   |__R/
      |__zzz.R
   |__README.md                    # this file
```




### 2. GWAS Data

----

As quoted and paraphrased from GWAS's [website](https://www.ebi.ac.uk/gwas/docs/about):
The Genome Wide Association Studies Catalog (denoted as GWAS) was founded by the [NHGRI](https://www.genome.gov/) in 2008. The GWAS Catalog provides a consistent, searchable, visualisable and freely available database of published SNP-trait associations (i.e., the association between common Single Nucleotide Variants and complex disease).
Within the Catalog, all eligible GWA studies are identified by literature search and assessed by the GWAS Catalog's curators, who then extract the reported trait, significant SNP-trait associations, and sample metadata.
The database curates eligible studies within 1-2 months of publication, and the data is released on a weekly cycle.
Since 2010, delivery and development of the Catalog has been a collaborative project between the EMBL-EBI and NHGRI. The Catalog website and infrastructure is now hosted by EMBL-EBI.

There are several versions available to download, as described [here](https://www.ebi.ac.uk/gwas/docs/fileheaders).
This package uses the 1.0.2 version, which is the most updated one, that also includes the ontology for the mapped trait, the study accession ID and the genotyping technology data
<br><br><br>

####**2.1 Data Semantics**  
Quoted and paraphrased from [GWAS websit](https://www.ebi.ac.uk/gwas/docs/file-downloads)
<br>
<br>

#####*File headers for catalog version 1.0.2*

- **DATE ADDED TO CATALOG**: Date a study is published in the catalog
- **PUBMEDID**: PubMed identification number
- **FIRST AUTHOR**: Last name and initials of first author
- **DATE**: Publication date (online (epub) date if available)
- **JOURNAL**: Abbreviated journal name
- **LINK**: PubMed URL
- **STUDY**: Title of paper
- **DISEASE/TRAIT**: Disease or trait examined in study
- **INITIAL SAMPLE DESCRIPTION**: Sample size and ancestry description for stage 1 of GWAS (summing across multiple Stage 1 populations, if applicable)
- **REPLICATION SAMPLE DESCRIPTION**: Sample size and ancestry description for subsequent replication(s) (summing across multiple populations, if applicable)
- **REGION**: Cytogenetic region associated with rs number
- **CHR_ID**: Chromosome number associated with rs number
- **CHR_POS**: Chromosomal position associated with rs number
- **REPORTED GENE(S)**: Gene(s) reported by author, HGNC symbols
- **MAPPED GENE(S)**: Gene(s) mapped to the strongest SNP, HGNC symbols. If the SNP is located within a gene, that gene is listed. If the SNP is intergenic, the upstream and downstream genes are listed, separated by a hyphen. Genes that are not charachterized and has no gene names, are listed as "LOC" and identifier number (according to [NCBI book](https://www.ncbi.nlm.nih.gov/books/NBK3840/#genefaq.Conventions).
- **UPSTREAM_GENE_ID**: Entrez Gene ID for nearest upstream gene to rs number, if not within gene
- **DOWNSTREAM_GENE_ID**: Entrez Gene ID for nearest downstream gene to rs number, if not within gene
- **SNP_GENE_IDS**: Entrez Gene ID, if rs number within gene; multiple genes denotes overlapping transcripts
- **UPSTREAM_GENE_DISTANCE**: distance in kb for nearest upstream gene to rs number, if not within gene
- **DOWNSTREAM_GENE_DISTANCE**: distance in kb for nearest downstream gene to rs number, if not within gene
- **STRONGEST SNP-RISK ALLELE**: SNP(s) most strongly associated with trait + risk allele (? for unknown risk allele). May also refer to a haplotype.
- **SNPS**: Strongest SNP; if a haplotype it may include more than one rs number (multiple SNPs comprising the haplotype)
- **MERGED**: denotes whether the SNP has been merged into a subsequent rs record (0 = no; 1 = yes;)
- **SNP_ID_CURRENT**: current rs number (will differ from strongest SNP when merged = 1)
- **CONTEXT**: SNP functional class
- **INTERGENIC**: denotes whether SNP is in intergenic region (0 = no; 1 = yes)
- **RISK ALLELE FREQUENCY**: Reported risk/effect allele frequency associated with strongest SNP in controls (if not available among all controls, among the control group with the largest sample size). If the associated locus is a haplotype the haplotype frequency will be extracted.
- **P-VALUE**: Reported p-value for strongest SNP risk allele (linked to dbGaP Association Browser). Note that p-values are rounded to 1 significant digit (for example, a published p-value of 4.8 x 10-7 is rounded to 5 x 10-7).
- **PVALUE_MLOG**: -log(p-value)
- **P-VALUE (TEXT)**: Information describing context of p-value (e.g. females, smokers).
- **OR or BETA**: Reported odds ratio or beta-coefficient associated with strongest SNP risk allele. Note that if an OR <1 is reported this is inverted, along with the reported allele, so that all ORs included in the Catalog are >1. Appropriate unit and increase/decrease are included for beta coefficients.
- **95% CI (TEXT)**: Reported 95% confidence interval associated with strongest SNP risk allele, along with unit in the case of beta-coefficients. If 95% CIs are not published, we estimate these using the standard error, where available.
- **PLATFORM (SNPS PASSING QC)**: Genotyping platform manufacturer used in Stage 1; also includes notation of pooled DNA study design or imputation of SNPs, where applicable
- **CNV**: Study of copy number variation (yes/no)
- **MAPPED_TRAIT**: Mapped Experimental Factor Ontology trait for this study
- **MAPPED_TRAIT_URI**: URI of the EFO trait
- **STUDY ACCESSION**: Accession ID allocated to a GWAS Catalog study
- **GENOTYPING_TECHNOLOGY**: Genotyping technology/ies used in this study, with additional array information (ex. Immunochip or Exome array) in brackets.

<br>
[**Full Description of Abbreviations used in the Catalog**](https://www.ebi.ac.uk/gwas/docs/abbreviations)
<br><br><br>


####**In summary:**
 - Association evidence in GWAS is determined according to the SNPs (or haplotypes) that are most strongly associated with trait + risk/effect allele, according to their reported P value. 
 - For each association, odds ratio (OR) or beta-coefficient (BETA) associated with strongest SNP risk allele are reported, as well as the risk allele frequency in controls (if not available among all controls, among the control group with the largest sample size)
 - Within the GWAS dataset, the column "MAPPED_GENE" provides the HGNC symbols of which the SNPs are mapped to. Some SNPs do not correspond to genes (i.e., intergenic) and some correspond to genes that were not yet characterized with HGNC symbols (denoted as "LOC" + identifier). This means that the main challenge for aligning between GWAS database and the appropriate HGNC symbols is to filter out the redundant SNP data, i.e., all SNPs that are not mapped to a gene that has a defined HGNC symbol. Another challenge would be to organize the filtered GWAS database to be comfortably accessible according to a desired HGNC symbol.
<br>
<br>
<br>

###**3 Data download and cleanup**

----


####**3.1 Preparations: packages**

 - To begin, we need to make sure the required packages are installed.
 - The following code was written by the author Boris Steipe and published at [GitHub](https://github.com/hyginn/R_Exercise-BasicSetup/blob/master/plottingIntro.R)  
<br>
<br>

According to [cran](https://cran.r-project.org/web/packages/data.table/index.html), **data.table** package provides functions to fast aggregate large data (e.g. 100GB in RAM), fast ordered joins, fast add/modify/delete of columns by group using no copies at all, list columns, friendly and fast character-separated-value read/write.

```R
if (!require(data.table, quietly=TRUE)){
  install.packages("data.table")
}
```
<br>

According to [cran](https://cran.r-project.org/web/packages/tidyr/index.html), **tidyr** package is designed specifically for data tidying (not general reshaping or aggregating) and works well with 'dplyr' data pipelines.

```R
if (!require(tidyr, quietly=TRUE)) {
  install.packages("tidyr")
}
```

<br><br>
 
####**3.2 Downloading the data**

**Notes**:


 - The data is over 57 megabytes size.
 - The API links to the most recent version release of the GWAS database. This database **updates** weekly, with the most recent update available at the same link.
 - A new **version** (not an update) may be released and published in a different URL. Caution needs to be taken: before using the following API, the user needs to check [GWAS downloads webpage](https://www.ebi.ac.uk/gwas/docs/file-downloads) for new releases and updates, and change the current used URL accordingly.
 - The code "as.data.frame(data.table::fread())" was written by the author Tarun Bhavnani and published at [Stacksoverflow](https://stackoverflow.com/questions/33322248/how-to-import-tsv-file-in-r), as a way to download tsv file according to an API.

<br>

Downloading the data according to the API provided in the GWAS website
```R
GWASraw <- as.data.frame(data.table::fread("https://www.ebi.ac.uk/gwas/api/search/downloads/full"))
```
<br>
<br>

What is the data's structure?
```R
str(GWASraw)
#'data.frame':	104767 obs. of  34 variables:
# $ DATE ADDED TO CATALOG     : chr  "2015-12-18" "2015-12-18" "2015-12-18" "2015-12-18" ...
# $ PUBMEDID                  : int  25574825 25574825 25574825 25574825 25574825 25574825 25574825 25574825 25574825 25574825 ...
# $ FIRST AUTHOR              : chr  "Baurecht H" "Baurecht H" "Baurecht H" "Baurecht H" ...
# $ DATE                      : chr  "2015-01-08" "2015-01-08" "2015-01-08" "2015-01-08" ...
# $ JOURNAL                   : chr  "Am J Hum Genet" "Am J Hum Genet" "Am J Hum Genet" "Am J Hum Genet" ...
# $ LINK                      : chr  "www.ncbi.nlm.nih.gov/pubmed/25574825" "www.ncbi.nlm.nih.gov/pubmed/25574825" "www.ncbi.nlm.nih.gov/pubmed/25574825" "www.ncbi.nlm.nih.gov/pubmed/25574825" ...
# $ STUDY                     : chr  "Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms." "Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms." "Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms." "Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms." ...
# $ DISEASE/TRAIT             : chr  "Inflammatory skin disease" "Inflammatory skin disease" "Inflammatory skin disease" "Inflammatory skin disease" ...
# $ INITIAL SAMPLE SIZE       : chr  "2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls" "2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls" "2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls" "2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls" ...
# $ REPLICATION SAMPLE SIZE   : chr  NA NA NA NA ...
# $ REGION                    : chr  "3q26.32" "2p25.3" "17q25.3" "10q26.3" ...
# $ CHR_ID                    : chr  "3" "2" "17" "10" ...
# $ CHR_POS                   : chr  "177009491" "2221629" "78284479" "129683009" ...
# $ REPORTED GENE(S)          : chr  "TBL1XR1" "MYT1L" "RP11-219G17.4" "MGMT" ...
# $ MAPPED_GENE               : chr  "LINC01209 - MTND5P15" "MYT1L" "LOC100996291 - LOC105371912" "MGMT" ...
# $ UPSTREAM_GENE_ID          : int  101928684 NA 100996291 NA 127623 106481095 353143 NA NA NA ...
# $ DOWNSTREAM_GENE_ID        : int  100873259 NA 105371912 NA 441932 400957 353142 NA NA NA ...
# $ SNP_GENE_IDS              : chr  "" "23040" "" "4255" ...
# $ UPSTREAM_GENE_DISTANCE    : int  192490 NA 5987 NA 38128 13503 5707 NA NA NA ...
# $ DOWNSTREAM_GENE_DISTANCE  : int  928 NA 57605 NA 958 2328 3029 NA NA NA ...
# $ STRONGEST SNP-RISK ALLELE : chr  "rs1384974-?" "rs6548123-?" "rs9302874-?" "rs80312298-?" ...
# $ SNPS                      : chr  "rs1384974" "rs6548123" "rs9302874" "rs80312298" ...
# $ MERGED                    : int  0 0 0 0 0 0 0 0 0 0 ...
# $ SNP_ID_CURRENT            : chr  "1384974" "6548123" "9302874" "80312298" ...
# $ CONTEXT                   : chr  "intergenic_variant" "intron_variant" "regulatory_region_variant" "intron_variant" ...
# $ INTERGENIC                : int  1 0 1 0 1 1 1 0 0 0 ...
# $ RISK ALLELE FREQUENCY     : chr  "NR" "NR" "NR" "NR" ...
# $ P-VALUE                   : chr  "4E-8" "6E-8" "2E-7" "6E-7" ...
# $ PVALUE_MLOG               : num  7.4 7.22 6.7 6.22 5.22 ...
# $ P-VALUE (TEXT)            : chr  "(Opposed)" "(Opposed)" "(Opposed)" "(Opposed)" ...
# $ OR or BETA                : num  NA NA NA NA NA 1.2 1.22 1.27 1.27 1.29 ...
# $ 95% CI (TEXT)             : chr  "" "" "" "" ...
# $ PLATFORM [SNPS PASSING QC]: chr  "Affymetrix, Illumina [~ 5200000]" "Affymetrix, Illumina [~ 5200000]" "Affymetrix, Illumina [~ 5200000]" "Affymetrix, Illumina [~ 5200000]" ...
# $ CNV                       : chr  "N" "N" "N" "N" ...
```
<br>

####**3.3 choosing the relevant columns**
As shown above, the raw GWAS database has 34 columns.  
Most columns hold data that does not concern to SNPs and their associated diseases and risks. The full semantics is described in chapter 2 of this README file.  
<br>
Leaving too many columns might cause confusion when trying to extract infrormation from the database, and even worse - the database would be too large to analyse due too increased size.  
<br>
Including columns that contain the associated disease, the mapped HGNC symbols, the SNP's identifiers ("SNPS", "MERGED", "SNP_ID_CURRENT"), the risk allele, the statistics of the association, the information whether the SNP was mapped to a gene or is intergenic:
```R
GWASprocessed <- GWASraw[,c("DISEASE/TRAIT", #of which the SNP has association to
                            "MAPPED_GENE", #The HGNC symbol of the gene that the SNP is mapped to
                            "STRONGEST SNP-RISK ALLELE", 
                            "SNPS", 
                            "MERGED", 
                            "SNP_ID_CURRENT",
                            "CONTEXT", 
                            "INTERGENIC", 
                            "RISK ALLELE FREQUENCY", 
                            "P-VALUE", 
                            "PVALUE_MLOG", 
                            "OR or BETA", 
                            "95% CI (TEXT)")] 
```
<br>
<br>

####**3.4 Filtering out unnecessary observations**
As explained earlier, there are non informative observations in the data that we need to filter out:
 - SNPs that were not mapped to HGNC symbols: intergenic SNPs, genes that were not characterized yet, missing data.
<br>
When observing the available observations along the "intergenic" column, we can see that besides 0 and 1, some rows retrieve NA:
```R
unique(GWASprocessed$INTERGENIC)
#[1]  1  0 NA

#How many NAs
sum(is.na(GWASprocessed$INTERGENIC))
#[1] 437
```
<br>
This means that filtering out SNPs that are intergenic can't be acheived with an expression such as `GWASprocessed<- GWASprocessed[GWASprocessed$INTERGENIC==0, ]`, which will retrieve rows of NAs whenever the "Intergenic" is NA.
<br>

When viewing those observations
```R
View(GWASprocessed[which(is.na(GWASprocessed$INTERGENIC)),])
```

####**3.4 Filtering out intergenic SNPs**
The "intergenic" column provides the information of whether the SNP is intergenic or within a gene.  
Our goal is to map SNPs to their corresponding HGNC symbols, which exists for genes and not for intergenic regions.
```R
GWASprocessed<- GWASprocessed[GWASprocessed$INTERGENIC==0, ]
```
<br>

Now, there is no further need for the "Intergenic" column, so we will remove it
```R
GWASprocessed$INTERGENIC <- NULL
```

<br><br>

####**3.5 Filtering out every SNP that isn't mapped to a gene**
For SNPs that were not mapped to a gene, the content of the "MAPPED_GENE" column is "" (I observed that fact when I searched for the optional blank values that I get as outputs when typing base::unique(GWASprocessed$MAPPED_GENE)
```R
GWASprocessed <- GWASprocessed[!GWASprocessed$MAPPED_GENE=="", ]
```
<br><br>

####**3.6 Separating rows of SNPs that has more than one mapped gene**
<br>
Separating the content of rows that has more than one value in the columns "MAPPED_GENE" and "SNP_GENE_IDS", so that each row has one value in the "MAPPED_GENE" column and it's corresponding "SNP_GENE_IDS" accession.
```R
GWASprocessed <- tidyr::separate_rows(data = GWASprocessed, 
                                      MAPPED_GENE, #Separating the rows with more than one
                                                   #Mapped gene
                                      SNP_GENE_IDS, #Separating the rows with more than
                                                    #one SNP gene ID
                                      sep = ", ") #The separator

```
<br><br>

####**3.7 Filtering out every SNP that is mapped to genes without characterization**
Mapped genes denoted with "LOC" + numeric identifier, correspond to genes that hadn't been characterized yet
Those genes have no corresponding HGNC symbols
```R
GWASprocessed <- GWASprocessed[!base::grepl(x = GWASprocessed$MAPPED_GENE,  pattern = "LOC"),]
```

                      
###**4.0 Mapping GWAS's SNP IDs to HGNC symbols**
----

<br>

 - Within the GWAS dataset, the column "MAPPED_GENE" provides the HGNC symbols of which the SNPs are mapped to.
 - Now, after cleaning the data to have only the SNPs that were mapped to genes with existing HGNC symbols, each HGNC symnbol can be mapped to it's own subset dataframe of SNPs.
 - In order to map a corresponding SNPs' data frame for each HGNC symbol, I will store each HGNC symbol as an individual list, that contains the data frame.
 - Note that HGNC symbols may changed with time (they are frequently updated), so the GWAS's MAPPED_GENE column might provide un updated HGNC symbol data.

<br><br>

####**4.1 **
```R
#Creating the vector of all HGNC symbols that appear in the GWAS database
HGNCsymbols <- base::unique(GWASprocessed$MAPPED_GENE)

#Creating a list of length of the number of HGNC symbols.
#The SNPs and association information of each HGNC symbol will be saved in each list.
HGNCtoGWAS <- base::vector("list", length = base::length(HGNCsymbols))

#A for loop that saves the data frame of the SNPs' information for each HGNC symbol that appears in GWAS database
for (i in base::seq(along = HGNCsymbols)){
HGNCpattern <- base::paste0("^", HGNCsymbols[i], "$") ##Setting the HGNC pattern to be
                                                      ##used to search for all relevant
                                                      ##Attributes with the grep function

#The location of the attributes of the specific HGNC symbol along the
#GWASprocessed data frame                                                      
HGNClocation <- base::grep(x = GWASprocessed$MAPPED_GENE, pattern = HGNCpattern)
                                                              
HGNCtoGWAS[[i]] <- GWASprocessed[HGNClocation, ] ##Adding the relevant attributes
                                                 ##to the list
                                                 
names(HGNCtoGWAS)[[i]] <- HGNCsymbols[i] ##Naming the list with the HGNC symbol
}
```
The HGNCtoGWAS list is a large list.


<br><br>

###**5.0 Validating**

<br><br>

###**6.0 Annotating example **
<br>

####**6.1 Loading the example gene set**

The example gene set was constructed by the author Boris Steipe, and copied from [BCB420-2019-resources repository](https://github.com/hyginn/BCB420-2019-resources/blob/master/exampleGeneSet.md)
```R
exampleGeneSet <- c("AMBRA1", "ATG14", "ATP2A1", "ATP2A2", "ATP2A3", "BECN1", "BECN2", 
"BIRC6", "BLOC1S1", "BLOC1S2", "BORCS5", "BORCS6", "BORCS7", 
"BORCS8", "CACNA1A", "CALCOCO2", "CTTN", "DCTN1", "EPG5", "GABARAP", 
"GABARAPL1", "GABARAPL2", "HDAC6", "HSPB8", "INPP5E", "IRGM", 
"KXD1", "LAMP1", "LAMP2", "LAMP3", "LAMP5", "MAP1LC3A", "MAP1LC3B", 
"MAP1LC3C", "MGRN1", "MYO1C", "MYO6", "NAPA", "NSF", "OPTN", 
"OSBPL1A", "PI4K2A", "PIK3C3", "PLEKHM1", "PSEN1", "RAB20", "RAB21", 
"RAB29", "RAB34", "RAB39A", "RAB7A", "RAB7B", "RPTOR", "RUBCN", 
"RUBCNL", "SNAP29", "SNAP47", "SNAPIN", "SPG11", "STX17", "STX6", 
"SYT7", "TARDBP", "TFEB", "TGM2", "TIFA", "TMEM175", "TOM1", 
"TPCN1", "TPCN2", "TPPP", "TXNIP", "UVRAG", "VAMP3", "VAMP7", 
"VAMP8", "VAPA", "VPS11", "VPS16", "VPS18", "VPS33A", "VPS39", 
"VPS41", "VTI1B", "YKT6")
```
<br><br>

####**6.2 Annotating the example gene set**

Checking how many out of of the example genes appear in the GWAS database
```R
#which of the example genes appear in GWAS
exampleGenesExsits <- exampleGeneSet[base::which(exampleGeneSet %in% names(HGNCtoGWAS))] 

#The percentage of existing example genes
exampleGenesExistsPerc <- (base::length(exampleGenesExsits) / base::length(exampleGeneSet)) *100
```
<br>

Annotating
```R

HGNCtoGWAS[exampleGenesExsits]
```


###Prerequesities
**Notes:**
 - In order to be able to execute the following codes, some required packages needs to be installed.
 - For that reason, whenever a package is required, the following code will be used named "RequiredPackage" must be loaded

**RequiredPackage()**
 - Installs the mentioned package if it isn't already installed
 - The code that is wrapped within the function was written and publised by:
 - The author: Boris Steipe
 - Published at: https://github.com/hyginn/R_Exercise-BasicSetup/blob/master/plottingIntro.R



**Notes:**
 - In order to be able to execute the following codes, some required packages needs to be installed.
 - For that reason, the following function named "RequiredPackage" must be loaded

**RequiredPackage()**
 - Installs the mentioned package if it isn't already installed
 - The code that is wrapped within the function was written and publised by:
 - The author: Boris Steipe
 - Published at: https://github.com/hyginn/R_Exercise-BasicSetup/blob/master/plottingIntro.R


*Description*
A function that checkes whether a package is already installed in R, and if it isn't installed, the function istalls it. 

*Usage*
RequiredPackage(PackageName)

*Arguments*
PackageName     A name of one package, with no wrapping "".

*Values*
This function does not produce values as outputs.
It downloads and installs the mentioned package if it was not already installed previously.

**The function**
```R
RequiredPackage <- function(PackageName){

  if (!require(PackageName, quietly=TRUE)) {
    install.packages("PackageName")
    }
}
```
 -----
#data.table is a required package for the following code
#The following condition ("if") installs the mentioned package if it isn't already installed.
#This "if" condition was written and publised by:
    #The author: Boris Steipe
    #Published at: https://github.com/hyginn/R_Exercise-BasicSetup/blob/master/plottingIntro.R
 -----
 #tidyr is a required package for the following code
#The following condition ("if") installs the mentioned package if it isn't already installed.
#This "if" condition was written and publised by:
    #The author: Boris Steipe
    #Published at: https://github.com/hyginn/R_Exercise-BasicSetup/blob/master/plottingIntro.R

#required package: tidyr
  if (!require(tidyr, quietly=TRUE)) {
    install.packages("tidyr")
    }
 -----
 



 

## 2.0 Prerequisites

You need a current installation of [**R**](https://www.r-project.org/) and [**RStudio**](https://www.rstudio.com/products/rstudio/download/), `git`, and a [**GitHub**](https://github.com/) account that **has been set up to connect to your RStudio projects**. If any of this is new to you (or if you wish to brush up on the details), head over to Jenny Bryan's superb tutorial [**Happy Git and GitHub with R**](http://happygitwithr.com/). You should also download the `devtools` and `testthat` packages from CRAN. In the RStudio console type:

```R
install.packages(c("devtools", "testthat"))
```

&nbsp;

## 2.1 A new GitHub project

Create a new, empty repository on GitHub and give it your package name.

- First you need to decide on a [**name**](http://r-pkgs.had.co.nz/package.html#naming) for your package. Take care to define it well. Short, memorable, lower-case, and not in conflict with current names on CRAN or Bioconductor. Head over to [the taskviews on CRAN](https://cran.r-project.org/web/views/) and browse to see good examples.
- Next, log into your GitHub account.
- Click on the **(+)** in the top menu bar and select _New repository_.
- Enter your package name as the repository name.
- _Check_ to **Initialize this repository with a README** (the README will be overwritten later, but you need at least one file as a placeholder in your repository.)
- Don't add a `.gitignore` file or a license (these will come from `rpt`).
- Click **Create repository**.
- Finally, copy the URL of your repository to your clipboard, it should look like `https://github.com/<your-GitHub-user-name>/<your-package-name>` <sup id="af2">[2](#f2)</sup>.

&nbsp;

## 2.2 A new RStudio project

Create a new RStudio project on your local machine that is linked to your GitHub repository and account.

- In RStudio, choose **File** ▷ **New Project...**, select **Version Control** ▷ **Git**. Enter the **Repository URL** you copied in the preceding step, hit your `tab` key to autofill the **Project directory name** (it should be the same as your package name), and **Browse...** to a parent directory in which you want to keep your RStudio project. Then click **Create Project**.

The project directory will be created, the repository file will be downloaded, a new RStudio session will open in your directory, and R's "working directory" should be set to here.

**Validate:**

1. In the console, type `getwd()`. This should print the correct directory.
2. Make a small change to the `README.md` file, commit it and push it back to the remote repository: 
    1. In the files pane, click on `README.md` to open the file in the editor. Make a small change (e.g. add the word "test"). Save the file.
    2. Click on the _Version control icon_ in the editor window menu and choose **Commit...**, or choose **Tools** ▷ **Version Control** ▷ **Commit...** from the menu.
    3. In the version control window, check the box next to `README.md` to "stage" the file, enter "test" as your "Commit message" and click **Commit**. This commits your edits to your local repository.
    4. Click the green **Push** up-arrow. This synchronizes your local repository with your remote repository on GitHub.
    5. Navigate to your GitHub repository, reload the page, and confirm that your edit has arrived in the `README.md` file in your GitHub repository.

Congratulate yourself if this has all worked. If not - don't continue. You need to fix whatever problem has arisen. In my experience, the most frequent issue is that someone has skipped a step that they thought was not important to them. Check carefully whether you have followed all the steps. In particular, if the problem is associated with `git` on your machine, or connecting RStudio to your GitHub repository, work through Jenny Bryan's [**Happy Git...**](http://happygitwithr.com/) first.

&nbsp;

## 2.3 Download the `rpt` files

Download a ZIP archive of `rpt` and copy all the files over to your project folder.

- Navigate to the GitHub repository for `rpt` at (<https://github.com/hyginn/rpt>).
- Click on the green **Clone or download** button and select **Download ZIP**. This will package the `rpt` folder into a ZIP archive which will contain all files, (without the actual repository database, you don't need that), and download it to your computer.
- Find the ZIP archive in your download folder and unpack it. This will create a folder called `rpt-master` which contains all of the `rpt` files. (Note:  the creation date of the folder is not today's date, so if your download folder lists files by date, the unzipped folder will not be at the top.)
- Move all of the files and folders within `rpt-master` into your project directory, overwriting any of the files that are already there. You can then delete `rpt-master` and the ZIP archive.

**Validate**

In RStudio, open the `./dev` directory. Open the file `rptTwee.R` and click on  **Source** to load the function. Then type `rptTwee()` into the console. You should get a directory tree that looks approximately like this.

```
 -- <your-package-name>/
   |__.gitignore
   |__.Rbuildignore
   |__DESCRIPTION
   |__dev/
      |__functionTemplate.R
      |__rptTwee.R
   |__inst/
      |__extdata/
         |__test_lseq.dat
      |__scripts/
         |__scriptTemplate.R
   |__LICENSE
   |__man/
      |__lseq.Rd
   |__NAMESPACE
   |__R/
      |__lseq.R
      |__zzz.R
   |__README.md
   |__<your-package-name>.Rproj
   |__rpt.Rproj
   |__tests/
      |__testthat.R
      |__testthat/
         |__helper-functions.R
         |__test_lseq.R
```

These are the files and directories for an RStudio project/R package that contains a sample function `lseq()` defined in `./R/lseq.R`, which is tested in `./test/testthat/test_lseq.R` using data in `./inst/extdata/test_lseq.dat`.

If directories or files are missing, figure out where you went wrong.

&nbsp;

## 2.4 Customize

Open the following `rpt` files in your script-editor, and edit/save them to customize them for your own package:

&nbsp;

#### `DESCRIPTION`

Modify the `DESCRIPTION` file as follows:

Replace `rpt` with your package name, change the version to 0.1.0 (a first development [version number](https://semver.org/)), change the date to today's date, remove the `cre` role from my `Authors@R` field, and change `aut` to `ctb`, add an author field for yourself, and write a description for your package.

```diff
-      Package: rpt
+      Package: <your-package-name>
Type: Package
-      Title: R Package Template
+      Title: <a title for your package>
-      Version: 1.2.2
+      Version: 0.1.0
-      Date: 2018-12-31
+      Date: <today-in-YYYY-MM-DD-format>
Authors@R: c(
-    person("Boris", "Steipe", email = "boris.steipe@utoronto.ca", role = c("aut", "cre"), comment = c(ORCID = "0000-0002-1134-6758"))
+     person("Boris", "Steipe", email = "boris.steipe@utoronto.ca", role = c("ctb"), comment = c(ORCID = "0000-0002-1134-6758")),
+     person("<your-given-name>", "<your-family-name>", email = "<your-email-address>", role = c("aut","cre"), comment = c(ORCID = "<your-ORCID-ID>"))
    )
-      Description: A template for an RStudio project of a basic R package ...
+      Description: {A short description of the purpose of your package}
License: file LICENSE
Encoding: UTF-8
LazyData: true
Suggests:
    testthat
RoxygenNote: 6.0.1

```
&nbsp;

### 2.4.1 Getting attribution right

Giving credit is the currency of the FOSS (Free and Open Source Software) world which makes all of our work possible, licensing keeps it free. Take the time to get your attributions and licenses right; even if you think you don't really need this immediately it's good practice for good habits. Don't think you don't have to care: you automatically have a copyright to everything you write, and if you don't license it, no one can legally re-use it. Unfortunately, the common practices for attributing R package authorship are not consistent wherever there is more than one author (which is usually the case in academia). `rpt` adopts a consistent approach that is backward compatible with earlier practice.

Attribution and licensing only appear to be related. They serve distinct requirements and require distinct and specific mechanisms.

Credible attribution needs to identify **who** authored **what** in a way that that information is conveniently accessible.

Credible licensing needs to identify **who** has a copyright to **what**, and under **which license** it is released, in a standard document.

- **Who**: all persons referenced in attributions or licenses - whether authors (`aut`), contributors (`ctb`), or other copyright holders (`cph`) - must be unambiguously identified. Since people's names are not unique, there is really only one good way to do this: associate everyone with their ORCID (Open Researcher and Contributor Identifier) ID. ORCID IDs are unique and stable. If you don't already have a (free!) [**ORCID ID**](https://orcid.org), now is a good time to get one - unless you don't identify as one who "participates in research, scholarship and innovation" at all. The common alternative of identifying persons by their e-mail is unique, but not stable. All authors and contributors are referenced in the `DESCRIPTION` file and the R packaging system uses standard methods to give credit. For details, in particular what the `aut` (author), `cre` (creator/maintainer), and `ctb` (contributor) roles mean, and which other fields might be important to you, see the [Package metadata chapter](http://r-pkgs.had.co.nz/description.html) in Hadley Wickham's book, and the [DESCRIPTION section](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#The-DESCRIPTION-file) of the CRAN "Writing R Extensions" manual.


- **What**: in any multi-author situation you need to specify exactly which files have been authored by whom. For **attribution**, add an `@author` tag to the Roxygen header of every source code item. For **licensing**, both the copyright and the licensed contents needs to be identified in the `LICENSE` file, and it must include the date (year), since copyright eventually expires. Yes, this is wordy and duplicates information. No, there's no obvious way to avoid that and still be compliant.

- **Which license**: There are many reasons to favour the MIT license over other FOSS licenses, `rpt` uses MIT. If you wish to use a different license, or need to include a different license because you are incorporating code that is differently licensed, then add the license, the contents it applies to, and the licensors' details into a clearly separated section of your LICENSE file.   

- **In practice**: I am the author (`aut`) and maintainer (`cre`) of the `rpt` package and this is reflected in the `DESCRIPTION` file. I have licensed `rpt` under the MIT license, and the MIT license requires that this information remains associated with the package. Therefore my information is listed both in the `DESCRIPTION` file, which feeds various mechanisms to document authorship, and the `LICENSE` file, which defines how others may modify, distribute and use the code. My package is however only a template for your own, you normally would not actually be _using_ the code I wrote. Therefore, the first thing you do is to add yourself as a `person`, and give yourself the `aut` and `cre` roles as author and maintainer, respectively. I am not a co-author even though I have contributed code initially: therefore my role changes to `ctb`. Over time you remove and replace my contributions with your own work, and at some point you can remove my attributions and copyright claims, while possibly adding attributions for other authors of code you use in your package, and collaborators. During this process, both the `DESCRIPTION` and the `LICENSE` file may contain more than one author and/or licensor. A common case is that you want to use a single function from a large package, or functions from a package that are not on CRAN. If this code is published under one of the FOSS licenses, you can simply copy the code, include it in your package, add the author to `Authors@R` - typically in a `ctb` role - and add their copyright information to the LICENSE file. Check the `DESCRIPTION` file, the `LICENSE` file, and the function headers for examples.

Now, having that considered, continue customising your files.

&nbsp;

#### `rpt.Rproj`

You already have a `<your-package-name>.Rproj` configuration file for RStudio in the main directory. You can either overwrite that with the options defined in `rpt.Rproj`, or set the options individually under **Tools** ▷ **Project options...** and delete `rpt.Rproj`. `rpt.Rproj` sets the following (significant) project options:

- A We **don't** save or restore the Workspace and we don't save History.<sup id="af3">[3](#f3)</sup> 
- B We use **two spaces** for indentation, not tabs.<sup id="af4">[4](#f4)</sup>
- C We use **UTF-8** encoding, always. There is no excuse not to.
- D The "BuildType" of the project is a **Package**. Once this is defined in the project options, the _Environment_ pane will include a tab for **Build** tools.

To implement these options:
- In the _Files_ pane, select `<your-package-name>.Rproj` and click on **Delete**.
- Select `rpt.Rproj` and **Rename** it to `<your-package-name>.Rproj`.
- Choose **File** ▷ **Recent Projects...** ▷ **&lt;your-package-name&gt;** and reload your project.

**Validate**

The _Environment_ pane should now have a **Build** tab.

&nbsp;

## 2.5 Save, check, commit, and push

It's time to complete the first development cycle: save, check, commit, and push to the `master` branch on GitHub.

1. **Save** all modified documents.
2. **Check** your package. Click on the **Build** tab, then click on the **Check** icon. This runs package checking code that confirms that all required files are present and correctly formatted, and all tests pass. See below.
3. Once your package check has passed without any errors, warnings or notes, click on the _Version control icon_ in the editor window menu and choose **Commit...**, or choose **Tools** ▷ **Version Control** ▷ **Commit...** from the menu.
4. In the version control window, check the box next to all changed files to "stage" them, enter "Initial Commit" as your "Commit message" and click **Commit**. 5. Click the green **Push** up-arrow to synchronize your local repository with GitHub.
6. Navigate to your GitHub repository, reload the page, and confirm that your edited files have arrived.


**Your package check must pass without errors, warnings or notes.** `rpt` passes the checks, and nothing you have done above should have changed this, if it was done correctly. Therefore something is not quite right if the checking code finds anything to complain about. Fix it now. You need a "known-good-state" to revert to, for debugging, in case problems arise later on.

**Validate**

Install your package from github and confirm that it can be loaded. In the console, type:

```R
devtools::install_github("<your-user-name>/<your-package-name>")
library(<your-package-name>)
citation("<your-package-name>")
?lseq
```

This should install your package, and load the library. Attaching the library runs the `.onAttach()` function in `./R/zzz.R` and displays the updated package name and authors.<sup id="af5">[5](#f5)</sup> The `citation()` function creates a package citation from information it finds in `DESCRIPTION`. It should include all author's names and the current year. The final command accesses the help page for the `lseq()` sample function that came with `rpt`, via R's help system. By confirming that this works, you are exercising functionality that is specific to the way R loads and manages packages and package metadata, none of which would work from information that has merely been left behind in your Workspace during development.

&nbsp;

# 3 Develop

You are done with configuring your baseline. **Check** your package frequently during development, and fix all errors right away. Package check errors have a way of interacting with each other that makes them hard to debug, it is best to address each one immediately when it occurs. Also, commit frequently and use meaningful commit messages. Your sanity will thank you. If you want to keep template files for reference, move them to the `./dev` directory so they will not be included in the package build. Finally, whenever you add new contents, reference it in the `LICENSE` file. Whenever you remove one of the original files, remove it from the `LICENSE` file. And whenever you modify a function, add your name to any existing authors.

&nbsp;

----
Some useful keyboard shortcuts for package authoring:

* Build and Reload Package:  `Cmd + Shift + B`
* Update Documentation:      `Cmd + Shift + D` or `devtools::document()`
* Test Package:              `Cmd + Shift + T`
* Check Package:             `Cmd + Shift + E` or `devtools::check()`

&nbsp;

# 4 What's in the box ...

Here is a list of assets provided with `rpt` and why they are included. You can delete everything you don't need, but note: you can't push empty directories to your repository. Make sure you keep at least one file in every directory that you want to keep during development.
 
```
.gitignore                     <- defines files that should not be committed to the repository
.Rbuildignore                  <- defines files that should not be included in the package
DESCRIPTION                    <- the metadata file for your package
dev                            <- optional: see (Note 1)
dev/functionTemplate.R         <- optional: see (Note 1)
dev/rptTwee.R                  <- optional: see (Note 1)
inst/                          <- optional: see (Note 2)
inst/extdata/                  <- optional: see (Note 3)
inst/extdata/test-lseq.dat     <- optional: see (Note 3)
inst/scripts/                  <- optional: see (Note 4)
inst/scripts/scriptTemplate.R  <- optional: see (Note 4)
LICENSE                        <- License(s)
man/                           <- help files, generated by Roxygen2: don't edit
NAMESPACE                      <- lists exported functions and data. Generated by Roxygen2: don't edit
R/                             <- Contains the code for exported functions
R/lseq.R                       <- a sample function
R/zzz.R                        <- three functions for package management
README.md                      <- see (Note 5)
rpt.Rproj                      <- project options. Rename to <your-package-name>.Rproj
tests                          <- see (Note 6)
tests/testthat                 <- contains scripts for tests to be run
tests/testthat/helper-functions.R  <- code runs to set up tests
tests/testthat/test_lseq.R     <- a test script for ./R/lseq.R
tests/testthat.R               <- the script that runs the tests
```

- **(Note 1)** The `./dev` directory. I use this directory to keep all files and assets that I need for development, but that should not be included and distributed in the final package. The directory is mentioned in `.Rbuildignore`. In `rpt` it contains `./dev/functionTemplate.R`, a template file for writing R functions with a Roxygen2 header, and `./dev/rptTwee.R`, which was discussed above.

- **(Note 2)** The `./inst` directory. Files in this directory are installed, and end up one level "higher" after installation. E.g. the contents of `./inst/extdata` is in the folder `./extdata/` of an installed package.

- **(Note 3)** The `./inst/extdata` directory. This directory commonly contains "extra" data that is used in tests and examples. (Actual package data would go into a top-level `./data` directory and needs to be "exported". See [the `rptPlus` package](https://github.com/hyginn/rptPlus) for an example.) Here it contains `inst/extdata/test-lseq.dat`, a sample data set used in the test for `lseq()`.

- **(Note 4)** The `./inst/scripts` directory. Many packages contain sample scripts in addition to the functions they share. Such  scripts go into this directory. `rpt` provides `./inst/scripts/scriptTemplate.R`, a template file to illustrate how to structure an R script.

- **(Note 5)** The file you are reading is the `README.md` file for the `rpt` package. `README` files explain what a package (or directory) contains, `.md` is the extension for [markdown](https://guides.github.com/features/mastering-markdown/) formatted text files. Replace the contents of this file with your own (you can keep using the [original on GitHub](https://github.com/hyginn/rpt/blob/master/README.md) as a reference); a nice template for structuring a markdown file is [here](https://gist.github.com/PurpleBooth/109311bb0361f32d87a2).

- **(Note 6)** The `./tests` directory contains directories and assets for tests. For details see the [**Testing**](http://r-pkgs.had.co.nz/tests.html) chapter of Hadley Wickham's book. 

&nbsp;

----

# 5 FAQ

##### How can I import Bioconductor packages?
Work with Bioconductor packages is described in the [`rptPlus`](https://github.com/hyginn/rptPlus) package template.

&nbsp;

# 6 Notes
- Syntax for footnotes in markdown documents was suggested by _Matteo_ [on Stackoverflow](https://stackoverflow.com/questions/25579868/how-to-add-footnotes-to-github-flavoured-markdown). (Regrettably, the links between footnote references and text don't work on GitHub.)

----

<b id="af1">1</b> A good way to begin your devevlopment journey is to first browse through [Hadley Wickham's book](http://r-pkgs.had.co.nz) to get an idea of the general layout of packages, then build a minimal package yourself, and then use the book, and the CRAN policies to hone and refine what you have done. You need a bit of knowledge to get you started, but after that, learning is most effective if you learn what you need in the context of applying it.  [↩](#a1).

<b id="af2">2</b> Empty repositories by convention have a `.git` extension to the repository name, repositories with contents have no extension: the name indicates the repository directory and that directory contains the `.git` file. Therefore your package should **NOT** be named `<package>.git` although links to your repository on GitHub seem to be correctly processed with both versions. For more discussion, see [here](https://stackoverflow.com/questions/11068576/why-do-some-repository-urls-end-in-git-while-others-dont) [↩](#a2)

<b id="af3">3</b> Among the R development "dogmas" that have been proven again and again by experience are:  "_Don't work in the console, always work in a script._" and "_Never restore old Workspace. Recreate your Workspace from a script instead._" Therefore my projects don't save history, and don't save (or restore) Workspace either. You don't have to follow this advice, but trust me: it's better practice. [↩](#a3)

<b id="af4">4</b> A commonly agreed on coding style is to use 80 character lines or shorter. That's often a bit of a challenge when you use spaces around operators, expressive variable names, and 4-space indents. Of those three, the 4-space indents are the most dispensable; using 2-space indents works great and helps keep lines short enough. There seems to be a recent trend towards 2-spaces anyway. As for tabs vs. spaces: I write a lot of code that is meant to be read and studied, thus I need more control over what my users see. Therefore I use spaces, not tabs. YMMV, change your Project Options if you feel differently about this. [↩](#a4)

<b id="af5">5</b> Displaying the startup message (as of this writing) works only once per session due to a long-standing bug in RStudio. (cf. [here](https://github.com/r-lib/devtools/issues/1442)). To display the message, choose **File** ▷ **Recent Projects...** ▷ **&lt;your-package-name&gt;** to reload your project, then type `library(<your-package-name>)` into the cosole. For more details on the namespace functions, see [here](https://stat.ethz.ch/R-manual/R-devel/library/base/html/ns-hooks.html).[↩](#a5)

&nbsp;

# 7 Further reading

- The [**R Packages** book](http://r-pkgs.had.co.nz/) 
- The [**CRAN** manual on writing R-extensions](https://cran.r-project.org/doc/manuals/r-release/R-exts.html)
- Hornik, K., Murdoch D. and Zeileis, A. (2012) [Who Did What? The Roles of R Package Authors and How to Refer to Them](https://journal.r-project.org/archive/2012-1/RJournal_2012-1_Hornik~et~al.pdf). The R Journal 4:(1) 64-69.
- [**roxygen2** documentation](https://cran.r-project.org/web/packages/roxygen2/vignettes/roxygen2.html)
- Cyril Chapellier's [**markdown cheatsheet**](https://github.com/tchapi/markdown-cheatsheet/blob/master/README.md) 
- The [**`rptPlus`** package template](https://github.com/hyginn/rptPlus) is built on `rpt` and covers more advanced topics.

&nbsp;

# 8 Acknowledgements

Thanks to my students, especially the BCB410 (Applied Bioinformatics) class of 2018, whose hard work on building their own R packages revealed the need for this template. [Yi Chen](https://orcid.org/0000-0003-1624-2760)'s careful proofreading helped make many points more specific. 

&nbsp;

&nbsp;


<a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.

<!-- END -->
